File: src/test/java/me/nasukhov/bot/BotTest.java
package me.nasukhov.bot;

import me.nasukhov.DI.ServiceLocator;
import me.nasukhov.bot.command.Handler;
import me.nasukhov.bot.io.*;
import me.nasukhov.bot.task.TaskManager;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.mockito.Mockito.*;

public class BotTest {
    Bot bot;
    Output output;
    ChannelRepository channelRepository;

    @BeforeEach
    void setup() {
        output = mock(Output.class);
        channelRepository = mock(ChannelRepository.class);
        bot = new Bot(channelRepository, mock(TaskManager.class));
        ServiceLocator.resetInstance();
        ServiceLocator locator = ServiceLocator.getInstance();
        locator.addDefinition(ChannelRepository.class, channelRepository);
    }

    @Test
    public void testHandleWhenChannelIsInactive(){
        Input input = new Input(
                "Unrelated input",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        when(channelRepository.findById(input.channel().id)).thenReturn(Optional.of(input.channel()));
        when(channelRepository.isActive(input.channel())).thenReturn(false);

        bot.handle(input, output);

        verify(channelRepository).findById(input.channel().id);
        verify(channelRepository).isActive(input.channel());
        verifyNoMoreInteractions(output, channelRepository);
    }

    @Test
    public void testHandleWhenNoHandlerIsPresent() {
        Input input = new Input(
                "Unrelated input",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        when(channelRepository.findById(input.channel().id)).thenReturn(Optional.of(input.channel()));
        when(channelRepository.isActive(input.channel())).thenReturn(true);

        bot.handle(input, output);

        verify(channelRepository).findById(input.channel().id);
        verify(channelRepository).isActive(input.channel());
        verifyNoMoreInteractions(output, channelRepository);
    }

    @Test
    public void testHandleWhenHandlerIsNotRegistered() {
        Input input = new Input(
                "/learn",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        when(channelRepository.findById(input.channel().id)).thenReturn(Optional.of(input.channel()));
        when(channelRepository.isActive(input.channel())).thenReturn(true);

        bot.handle(input, output);

        verify(channelRepository).findById(input.channel().id);
        verify(channelRepository).isActive(input.channel());
        verifyNoMoreInteractions(output, channelRepository);
    }

    @Test
    public void testHandle() {
        Input input = new Input(
                "/learn",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        Handler handler = mock(Handler.class);
        bot.addHandler(handler);

        when(channelRepository.findById(input.channel().id)).thenReturn(Optional.of(input.channel()));
        when(channelRepository.isActive(input.channel())).thenReturn(true);
        when(handler.supports(input)).thenReturn(true);

        bot.handle(input, output);

        verify(channelRepository).findById(input.channel().id);
        verify(channelRepository).isActive(input.channel());
        verify(handler).supports(input);
        verify(handler).handle(input, output);
        verifyNoInteractions(output);
    }

    @Test
    public void testHandleForNonRegisteredChannel() {
        Input input = new Input(
                "/learn",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        Handler handler = mock(Handler.class);
        bot.addHandler(handler);

        when(channelRepository.findById(input.channel().id)).thenReturn(Optional.empty());
        doNothing().when(channelRepository).saveChannel(input.channel());
        when(handler.supports(input)).thenReturn(true);

        bot.handle(input, output);

        verify(channelRepository).findById(input.channel().id);
        verify(channelRepository).saveChannel(input.channel());
        verify(handler).supports(input);
        verify(handler).handle(input, output);
        verifyNoInteractions(output);
    }
}

File: src/test/kotlin/me/nasukhov/bot/BotTest.kt
package me.nasukhov.bot

import me.nasukhov.bot.command.Handler
import me.nasukhov.bot.io.Channel
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.NoOutput
import me.nasukhov.bot.io.TextOutput
import me.nasukhov.bot.io.User
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertInstanceOf
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.springframework.context.ApplicationEventPublisher
import kotlin.test.assertSame

class BotTest {
    private lateinit var bot: Bot

    @BeforeEach
    fun setup() {
        bot =
            Bot(
                mutableListOf(),
                Mockito.mock(ApplicationEventPublisher::class.java),
            )
    }

    @Test
    fun testHandleWhenNoHandlerIsPresent() {
        val channel = Channel("SomeChannelId")
        val user = User("SomeUserId", "SomeUserName")
        val input = Input("Unrelated input", channel, user)

        val result = bot.handle(input)
        assertInstanceOf<NoOutput>(result)
    }

    @Test
    fun testHandleUnsupportedCommand() {
        val channel = Channel("SomeChannelId")
        val user = User("SomeUserId", "SomeUserName")
        val input = Input("Unrelated input", channel, user)

        val handler = Mockito.mock(Handler::class.java)
        Mockito.`when`(handler.supports(input)).thenReturn(false)
        bot.addHandler(handler)

        val result = bot.handle(input)
        assertInstanceOf<NoOutput>(result)

        Mockito
            .verify(handler, Mockito.never())
            .handle(any())
    }

    @Test
    fun testHandle() {
        val channel = Channel("SomeChannelId")
        val user = User("SomeUserId", "SomeUserName")
        val input = Input("Unrelated input", channel, user)

        val handler = Mockito.mock(Handler::class.java)
        Mockito.`when`(handler.supports(input)).thenReturn(true)
        val handlerOutput = TextOutput("Some output")
        Mockito.`when`(handler.handle(input)).thenReturn(handlerOutput)

        bot.addHandler(handler)

        val result = bot.handle(input)
        assertSame(handlerOutput, result)
    }
}
